c10/core/DispatchKey.h

```diff
@@ -360,54 +513,83 @@ enum class DispatchKey : uint8_t {
-      return 0;
-    case DispatchKey::CPU:
-      return 1;
-    case DispatchKey::QuantizedCPU:
-      return 2;
-    case DispatchKey::SparseCPU:
-      return 3;
-    case DispatchKey::BackendSelect:
-      return 4;
-    case DispatchKey::ADInplaceOrView:
-      return 5;
-    case DispatchKey::AutogradOther:
-      return 6;
-    case DispatchKey::AutogradCPU:
-      return 7;
-    case DispatchKey::NumDispatchKeys: // Sentinel, end of runtime keys.
-      return 8;
-    default:
-      return -1;
+// Check if a DispatchKey is an alias mapping to other runtime keys.
+constexpr bool isAliasDispatchKey(DispatchKey k) {
+  return k >= DispatchKey::StartOfAliasKeys && k <= DispatchKey::EndOfAliasKeys;
+}
+
+// [Note: Per-Backend Functionality Dispatch Keys]
+// Check if a DispatchKey is a per-backend functionality key
+// Any functionalities that can be customized per-backend should be added here.
+// These keys correspond to functionalities that can be customized indivually per backend.
+// While they only take up one bit in the `DispatchKeySet` bitset,
+// they map to (# backends) slots in the operator table.
+// Each of these keys also has a separate set of "runtime keys" in the dispatch key enum,
+// per backend, which *do* map to the individual operator table slots.
+// For example, the "Sparse" key maps to an individual bit in the DispatchKeySet,
+// while `SparseCPU`, `SparseCUDA`, etc all map to individual slots in the runtime operator table.
+
+constexpr bool isPerBackendFunctionalityKey(DispatchKey k) {
+  if (k == DispatchKey::Dense ||
+      k == DispatchKey::Quantized ||
+      k == DispatchKey::Sparse ||
+      k == DispatchKey::AutogradFunctionality) {
+      return true;
+  } else {
+      return false;
   }
 }
-#else
-/**
- * For the server use-case, make this a simple pass-through.
- */
-C10_API constexpr int getDispatchTableIndexForDispatchKey(DispatchKey dk) {
-  return static_cast<int>(dk);
+
+// Note that this includes Undefined in the total count.
+// BUT EndOfFunctionalityKeys is its own (placeholder) key.
+// e.g. Undefined=0, Dense=1, Sparse=2, EndOfFunctionalityKeys=3.
+// In the above example, there are 3 total functionality keys.
+constexpr uint8_t num_functionality_keys =
+    static_cast<uint8_t>(DispatchKey::EndOfFunctionalityKeys);
+
+// Note [No More Than 16 Backends]
+// Search for this note to find places in the code where the "no more than 16 backends"
+// invariant is baked in.
+static_assert(
+    static_cast<uint8_t>(BackendComponent::EndOfBackendKeys) <= 16,
+    "BackendComponent currently only supports <= 16 backends. If we really need to extend this, \
+there are a few places where this invariant is baked in");
+
+constexpr uint8_t numPerBackendFunctionalityKeys() {
+  uint8_t count = 0;
+  for (uint8_t k = 0; k <= num_functionality_keys; ++k) {
+    if (isPerBackendFunctionalityKey(static_cast<DispatchKey>(k))) ++count;
+  }
+  return count;
 }
+
+
+#if defined(C10_MOBILE_TRIM_DISPATCH_KEYS)
```

Comment:
Hey @dhruvbird, the dispatcher changes in this PR required some changes to the mobile-only code, let me know if you want to take a deeper look at it.

I tried building locally with `C10_MOBILE_TRIM_DISPATCH_KEYS=1` and making sure that basic functionality works, but other than that I'm relying on mobile CI (internal + OSS) to catch any issues. Since this is a pretty low-level change, are there any tests that you think I should explicitly kick off in sandcastle?